File: backend/internal/utils/slug.go
package utils

import (
	"crypto/sha256"
	"encoding/base64"
	"fmt"
	"path/filepath"
	"regexp"
	"strings"
)

// GenerateSlug creates a slug from filename and short hash.
func GenerateSlug(filename, content string) string {
	base := filepath.Base(filename)
	ext := filepath.Ext(base)
	name := base[:len(base)-len(ext)]

	re := regexp.MustCompile(`[^a-zA-Z0-9\-]`)
	name = strings.ToLower(name)
	name = strings.ReplaceAll(name, "_", "-")
	name = strings.ReplaceAll(name, " ", "-")
	name = re.ReplaceAllString(name, "")

	words := strings.Split(name, "-")
	if len(words) > 3 {
		words = words[:3]
	}

	name = strings.Join(words, "-")
	if len(name) > 20 {
		name = name[:20]
	}

	name = strings.TrimRight(name, "-")

	hash := sha256.Sum256([]byte(content))
	shortHash := base64.URLEncoding.EncodeToString(hash[:4])
	shortHash = strings.TrimRight(shortHash, "=")

	if name == "" {
		return shortHash
	}

	return fmt.Sprintf("%s--%s", name, shortHash)
}

File: cli/slug.ts
import { createHash } from 'node:crypto';
import path from 'node:path';

export function generateSlug(filename: string, content: string): string {
  const base = path.basename(filename);
  const ext = path.extname(base);
  let name = base.slice(0, base.length - ext.length);

  name = name.toLowerCase();
  name = name.replaceAll('_', '-').replaceAll(' ', '-');
  name = name.replace(/[^a-z0-9-]/g, '');

  const words = name.split('-').filter(Boolean);
  const shortWords = words.slice(0, 3);
  name = shortWords.join('-').slice(0, 20).replace(/-+$/g, '');

  const hash = createHash('sha256').update(content).digest();
  const shortHash = Buffer.from(hash.subarray(0, 4)).toString('base64url');

  if (!name) {
    return shortHash;
  }

  return `${name}--${shortHash}`;
}
